### PR TITLE
Fix add-implicit-clauses middleware not properly handling source queries

### PR DESCRIPTION
**Pull request by camsaul**
Split off from #19384 

The `add-implicit-clauses` middleware did not properly transform source queries. Fixes the bug and adds some tests

### CODE REVIEW

File: src/metabase/query_processor/middleware/add_implicit_clauses.clj [hunk 1]
```diff
@@ -125,14 +134,17 @@
...
-      ;; if query has an MBQL source query recursively add implicit clauses to that too as needed
-      (update inner-query :source-query add-implicit-mbql-clauses)
-      ;; otherwise we're done
-      inner-query)))
+  [form]
+  (walk/postwalk
+   (fn [form]
+     ;; add implicit clauses to any 'inner query', except for joins themselves (we should still add implicit clauses
+     ;; like `:fields` to source queries *inside* joins)
+     (if (and (map? form)
```
Comment (howonlee) [inline src/metabase/query_processor/middleware/add_implicit_clauses.clj:142]: it seems that this postwalk will just shove the implicit clauses even into the native queries now? was that intended?

Comment (camsaul) [reply src/metabase/query_processor/middleware/add_implicit_clauses.clj:142]: see the line below: it only transforms maps that have `:source-table` or `:source-query` in them -- that's how we know it's an MBQL query and not a native query

**Review by howonlee — APPROVED** [review]
I was a bit confused at explicit prevention of this applying to native queries being gone but there's an implicit whacking of native queries so it's fine

**Comment by camsaul** [comment]
Just going to merge this as part of #19384

**Pull request closed without merging**